dimod/higherorder.py

```diff
@@ -23,8 +23,92 @@
 from dimod.binary_quadratic_model import BinaryQuadraticModel
 from dimod.decorators import vartype_argument
 from dimod.vartypes import Vartype
+import dimod
+from dimod.response import SampleSet
+import numpy as np
 
-__all__ = ['make_quadratic']
+__all__ = ['make_quadratic','HigherOrderComposite']
+
+
+class HigherOrderComposite(dimod.ComposedSampler):
+
+    def __init__(self, child_sampler):
+        self._children = [child_sampler]
+
+    @property
+    def children(self):
+        return self._children
+
+    @property
+    def parameters(self):
+        param = self.child.parameters.copy()
+        param['penalty_strength'] = []
+        return param
+
+    @property
+    def properties(self):
+        return {'child_properties': self.child.properties.copy()}
+
+    def sample(self, h, J, offset=0, penalty_strength=1.0, **parameters):
+        bqm = BinaryQuadraticModel(linear=h, quadratic={}, offset=0,
+                                   vartype=dimod.SPIN)
+        bqm = make_quadratic(J, penalty_strength, bqm=bqm)
+        response = self.child.sample(bqm, **parameters)
+
+        return polymorph_response(response, h, J, offset, bqm,
+                                  penalty_strength=penalty_strength)
+
+
+def polymorph_response(response, h, J, offset, bqm, penalty_strength=None):
+    record = response.record
+    original_variables = set(
+        sorted(set(h.keys()) | set(v for key in J.keys() for
+                                   v in key)))
+
+    penalty_vector = check_penalty_satisfaction(response, bqm)
+
+    poly = _shifted_poly(h, J, response.label_to_idx)
+    energy_vector = np.add(poly_energy(record.sample, poly), offset)
+    idxs = [response.label_to_idx[v] for v in original_variables]
+    samples = np.asarray(record.sample[:, idxs])
+    num_samples, num_variables = np.shape(samples)
+
+    datatypes = [('sample', np.dtype(np.int8), (num_variables,)),
+                 ('energy', energy_vector.dtype),
+                 ('penalty_satisfaction',
+                  penalty_vector.dtype)]
+
+    datatypes.extend((name, record[name].dtype, record[name].shape[1:])
+                     for name in record.dtype.names if
+                     name not in {'sample',
+                                  'energy'})
+
+    data = np.rec.array(np.empty(num_samples, dtype=datatypes))
+
+    data.sample = samples
+    data.energy = energy_vector
+    for name in record.dtype.names:
+        if name not in {'sample', 'energy'}:
+            data[name] = record[name]
+
+    data['penalty_satisfaction'] = penalty_vector
+    response.info['reduction'] = bqm.info['reduction']
+    response.info['penalty_stength'] = penalty_strength
+    return SampleSet(data, original_variables, response.info,
+                     response.vartype)
+
+
+def check_penalty_satisfaction(response, bqm):
```

Comment:
I think this is a misleading function name because it returns a vector. I would expect check to return a bool or throw an exception.